x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React, { useState, useEffect, useRef } from 'react';
+import { isEqual } from 'lodash';
+import {
+  Map,
+  MapboxOptions,
+  NavigationControl,
+  Popup,
+  PopupOptions,
+  GeoJSONSourceOptions
+} from 'mapbox-gl';
+
+interface Props {
+  style?: React.CSSProperties;
+  getColorStyle: (scale: number) => string;
+  mapboxStyle: MapboxOptions['style'];
+  geojsonSource: NonNullable<GeoJSONSourceOptions['data']>;
+  geojsonKeyProperty: string;
+  data: Array<{ key: string; value: number; [property: string]: any }>;
+  maxValue: number;
+  renderTooltip: (props: {
+    geojsonProperties: NonNullable<mapboxgl.MapboxGeoJSONFeature['properties']>;
+    data?: Props['data'][0];
+  }) => React.ReactElement | null;
+  initialMapboxOptions?: Partial<MapboxOptions>;
+}
+export const ChoroplethMap: React.SFC<Props> = props => {
+  const {
+    style,
+    getColorStyle,
+    mapboxStyle = 'https://tiles.maps.elastic.co/styles/osm-bright/style.json',
+    geojsonSource,
+    geojsonKeyProperty,
+    data,
+    maxValue,
+    renderTooltip: ToolTip
+  } = props;
+
+  const [hoveredData, setHoveredData] = useState<Props['data'][0] | undefined>(
+    undefined
+  );
+  const [hoveredFeatureProperties, setHoveredFeatureProperties] = useState<
+    mapboxgl.MapboxGeoJSONFeature['properties']
+  >(null);
+  const hoveredFeaturePropertiesRef = useRef<
+    mapboxgl.MapboxGeoJSONFeature['properties']
+  >(null);
+  const isMapReady = useRef<boolean>(false);
```

Comment:
Could you explain why you're using `useRef` instead of `useState`/`useCallback`? I've got a feeling this code would be simpler and more robust if we avoid useRef.